### PR TITLE
 Move DNS name mapping from endpoint to report

### LINKED ISSUE #3058: Move dns names from endpoints to report level
In large reports we can see many repeats of the same name lookup:

```
      ";10.244.252.166;42230": {
        "id": ";10.244.252.166;42230",
        "topology": "endpoint",
        "counters": null,
        "sets": {
          "reverse_dns_names": [
            "ip-10-244-252-166.ec2.internal"
          ]
        },
...
      ";10.244.252.166;42170": {
        "id": ";10.244.252.166;42170",
        "topology": "endpoint",
        "counters": null,
        "sets": {
          "reverse_dns_names": [
            "ip-10-244-252-166.ec2.internal"
          ]
        },
...
      ";10.244.252.166;42200": {
        "id": ";10.244.252.166;42200",
        "topology": "endpoint",
        "counters": null,
        "sets": {
          "reverse_dns_names": [
            "ip-10-244-252-166.ec2.internal"
          ]
        },
...
```

so we could have a structure up at `Report` level that mapped them once, then just look this up when we need it.
For backwards-compatibility we could create this structure in `Report.Upgrade()`

Relates to #1201

### PR DESCRIPTION
**Pull request by bboreham**
Fixes #3058 

If the probe sees many TCP sockets on a host, it will repeat the same address information on each endpoint.  Moving it to report level saves all this repetition.  Also `Sets` is an expensive data structure; for the DNS mapping I used a straightforward `map`, similar to `Topology.Nodes`.

I see about a 6% reduction in JSON file size, on Weave Cloud production reports.

The mapping is used as a mapping from IP address to name, and the current implementation encodes it the same way, but to me it seems more logical to send it as a mapping from name to address.  Probably that doesn't matter.

~I considered moving `DNSFirstMatch` from `render` to `report`, to collect all the behaviour together.  However the decisions made are somewhat related to rendering, so I left it.~

### CODE REVIEW

**Comment by rade** [comment]
This is supposed to fix #3058, right?

File: render/detailed/connections.go [hunk 1]
```diff
@@ -73,7 +73,7 @@ func newConnectionCounters() *connectionCounters {
 	return &connectionCounters{counted: map[string]struct{}{}, counts: map[connection]int{}}
 }
 
-func (c *connectionCounters) add(outgoing bool, localNode, remoteNode, localEndpoint, remoteEndpoint report.Node) {
+func (c *connectionCounters) add(rpt report.Report, outgoing bool, localNode, remoteNode, localEndpoint, remoteEndpoint report.Node) {
```
Comment (rade) [inline render/detailed/connections.go]: perhaps this should take just a `DNSRecords` instead of a whole `Report`.

**Comment by bboreham** [comment]
@rade yes, I forgot I'd filed an issue.

File: report/report.go [hunk 2]
```diff
@@ -469,6 +475,37 @@ func (r Report) upgradeNamespaces() Report {
 	return r
 }
 
+// Note in-place modification of Endpoint.Nodes
```
Comment (rade) [inline report/report.go]: This is way dodgy. Don't do it. If you must, I expect an essay here explaining why it is safe.

File: report/report.go [hunk 3]
```diff
@@ -469,6 +475,37 @@ func (r Report) upgradeNamespaces() Report {
...
+		return r
+	}
+	dns := make(DNSRecords)
+	for endpointID, endpoint := range r.Endpoint.Nodes {
+		_, addr, _, ok := ParseEndpointNodeID(endpointID)
+		snoopedNames, foundS := endpoint.Sets.Lookup(SnoopedDNSNames)
+		reverseNames, foundR := endpoint.Sets.Lookup(ReverseDNSNames)
+		if ok && (foundS || foundR) {
+			// Add address and names to report-level data - we assume
+			// all endpoints with the same address have equally good data
```
Comment (rade) [inline report/report.go]: I don't see why that assumption would hold. And if it doesn't then the data here does not converge, which results in view instability.

You could _optimise_ for the case that the assumption holds, but not depend on it.

File: report/report.go (hunk 2 above)
Comment (bboreham) [reply report/report.go]: I was under the impression the existing `upgrade` routines did it too, but on closer inspection they don't.  Will fix.

File: report/report.go (hunk 3 above)
Comment (bboreham) [reply report/report.go]: The data comes from the caches in `resolver.go` and `dns_snooper_linux.go`; if that data is changing then it's a matter of timing what any particular endpoint has.

I can't imagine why the data would change in such a way that results didn't converge shortly.

Comment (rade) [reply report/report.go]: AFAICT you could process the same report twice and get different results. That's a recipe for view instability.

**Comment by bboreham** [comment]
I made the requested changes and rebased.  Also moved `DNSFirstMatch` since it looks a lot more like a method on `DNSRecords` now.

**Pull request merged**